### PR TITLE
Fms2io implementation 

### LINKED ISSUE #7: Implement the new io
**Is your feature request related to a problem? Please describe.**
N/A

**Describe the solution you'd like**
FMS has a new io: fms2_io that should be used instead of fms_io

**Describe alternatives you've considered**
N/A

**Additional context**

### PR DESCRIPTION
**Pull request by uramirez8707**
**Description**
This PR implements the new io to read/write  `atmos_coupled.res.nc` and adds some changes required to implement the new io into atmos_params

1. Assigns the atmosphere domain to Physics%control%domain, so that the domain can be pass down and used in the physics when using the new io 
2. Created subroutines: `register_atmos_restart_scalar` and `register_atmos_restart_domain` to register the restart variables for `atmos_coupled.res.nc` and  'add_domain_dimension_data` to add dummy axis data, which is needed for the combiner to work correctly. 
3. Replaces ascii_io
4. Removes the if block that handles the "native formatted restart file", so netcdf files are the only file types supported :)
```F90
   else if (file_exist('INPUT/atmos_coupled.res')) then
          if(mpp_pe() == mpp_root_pe() ) call mpp_error ('atmos_model_mod', &
                   'Reading native formatted restart file: INPUT/atmos_coupled.res', NOTE)
           ...
```

Fixes #7 

**How Has This Been Tested?**
AM4 with intel18, debug/prod on GAEA.
doubly_periodic XML with intel18, prod, for 1 tile runs.
CM4/ESM4 runs with prod

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules
Requires [atmos_phys changes](https://gitlab.gfdl.noaa.gov/fms/atmos_phys/-/merge_requests/16) to work correctly. 
- [x] New check tests, if applicable, are included

### CODE REVIEW

**Comment by uramirez8707** [comment]
@menzel-gfdl

**Comment by thomas-robinson** [comment]
@mlee03 has the doubly_periodic XML that can run on a single tile.

**Review by wrongkindofdoctor — APPROVED** [review]

File: coupled/atmos_model.F90 [hunk 1]
```diff
@@ -771,19 +767,9 @@ subroutine atmos_model_init (Atmos, Time_init, Time, Time_step, &
              Cosp_rad(idx), &
              Moist_clouds(idx))
 
-   IF ( file_exist('input.nml')) THEN
-#ifdef INTERNAL_FILE_NML
+   IF ( file_exists('input.nml')) THEN
```
Comment (menzel-gfdl) [inline coupled/atmos_model.F90:770]: @uramirez8707 Do we need this if statement?

Comment (uramirez8707) [reply coupled/atmos_model.F90:770]: The file *should* exist by then.

Comment (menzel-gfdl) [reply coupled/atmos_model.F90:770]: Yes I think mpp would crash before this if the file isn't in the current directory.

**Comment by uramirez8707** [comment]
@thomas-robinson  This needs to be merged and tagged.

**Comment by thomas-robinson** [comment]
I assigned to @laurenchilutti for merging and tagging.  She is the primary manager for this repo.  Sorry this got lost in the shuffle

**Comment by laurenchilutti** [comment]
Is this okay to merge then? It appears to be waiting on approval from @thomas-robinson and @menzel-gfdl

File: coupled/atmos_model.F90 [hunk 2]
```diff
@@ -768,21 +767,11 @@ subroutine atmos_model_init (Atmos, Time_init, Time, Time_step, &
...
-      ierr=1
-      do while (ierr /= 0)
-         read  (unit, nml=atmos_model_nml, iostat=io, end=10)
-         ierr = check_nml_error(io,'atmos_model_nml')
-      enddo
- 10     call close_file (unit)
-#endif
    endif
-   call get_restart_io_mode(do_netcdf_restart)
+   do_netcdf_restart = .true. !< Always use netcdf
```
Comment (thomas-robinson) [inline coupled/atmos_model.F90:774]: This is inflexible for the future if it ever changes.

File: solo/atmos_model.F90 [hunk 3]
```diff
@@ -25,34 +25,24 @@ program atmos_model
...
+use          fms_mod, only: check_nml_error,                &
                             error_mesg, FATAL, WARNING,                 &
                             mpp_pe, mpp_root_pe, fms_init, fms_end,     &
                             stdlog, stdout, write_version_number,       &
-                            open_restart_file, lowercase,               &
+                            lowercase,               &
                             mpp_clock_id, mpp_clock_begin,              &
-                            mpp_clock_end, CLOCK_COMPONENT, set_domain, nullify_domain
+                            mpp_clock_end, CLOCK_COMPONENT
 use       fms_io_mod, only: fms_io_exit
```
Comment (thomas-robinson) [inline solo/atmos_model.F90:42]: Is this needed?

**Review by thomas-robinson — APPROVED** [review]

File: solo/atmos_model.F90 (hunk 3 above)
Comment (uramirez8707) [reply solo/atmos_model.F90:42]: `set_domain` and `nullify_domain` are fms_io functions that are no longer needed.

**Review by menzel-gfdl — APPROVED** [review]

File: coupled/atmos_model.F90 (hunk 2 above)
Comment (uramirez8707) [reply coupled/atmos_model.F90:774]: Then we can update/change that in the "future"

File: solo/atmos_model.F90 [hunk 4]
```diff
@@ -365,7 +350,7 @@ end subroutine atmos_model_init
 
    subroutine atmos_model_end
 
-   integer :: unit, date(6)
+   integer :: restat_file_unit, date(6)
```
Comment (bensonr) [inline solo/atmos_model.F90:353]: Think you have a typo here - should be "restart"

File: solo/atmos_model.F90 [hunk 5]
```diff
@@ -394,11 +379,10 @@ subroutine atmos_model_end
...
 
       if ( mpp_pe() == mpp_root_pe() ) then
-           call mpp_open (unit, 'RESTART/atmos_model.res', form=MPP_ASCII, action=MPP_OVERWR, &
-                          access=MPP_SEQUENTIAL, threading=MPP_SINGLE, nohdrs=.true. )
-           write (unit,'(6i6,8x,a)') date, &
+           open(newunit = restart_file_unit, file='RESTART/atmos_model.res', status='replace', form='formatted')
+           write (restart_file_unit,'(6i6,8x,a)') date, &
                  'Current model time: year, month, day, hour, minute, second'
-           call mpp_close (unit)
+           close (restat_file_unit)
```
Comment (bensonr) [inline solo/atmos_model.F90:385]: Think you have a typo here - should be "restart"

Comment (uramirez8707) [reply solo/atmos_model.F90]: Good catch. Fixed in https://github.com/NOAA-GFDL/atmos_drivers/pull/8/commits/45576ed4459bab5641c45e12ec86fd3e8fe1d021

File: solo/atmos_model.F90 (hunk 3 above)
Comment (thomas-robinson) [reply solo/atmos_model.F90:42]: So if you deleted them, then do you need fms_io_exit?

Comment (uramirez8707) [reply solo/atmos_model.F90:42]: fms_io_exit is where the writes actually happen, so if fms_io is used somewhere else, removing this call will break it (probably)
https://github.com/NOAA-GFDL/FMS/blob/b9fc6515c7e729909e59a0f9a1efc6eb1d3e44d1/fms/fms_io.F90#L44-L46
https://github.com/NOAA-GFDL/FMS/blob/main/fms/fms_io.F90#L778-L913

so this call should stay until fms_io is completely removed.

**Comment by laurenchilutti** [comment]
@bensonr do you approve of Uriel's latest commit?  If so I can go ahead and merge this.

**Review by bensonr — APPROVED** [review]

**Pull request merged**